lib/ts/recipe/totp/recipe.ts

```diff
@@ -0,0 +1,243 @@
+                pathWithoutApiBasePath: new NormalisedURLPath(LIST_DEVICE_API),
+            },
+            {
+                id: IS_TOTP_ENABLED_API,
+                disabled: this.apiImpl.isTotpEnabledGET === undefined,
+                method: "get",
+                pathWithoutApiBasePath: new NormalisedURLPath(IS_TOTP_ENABLED_API),
+            },
+        ];
+    };
+
+    handleAPIRequest = async (
+        id: string,
+        req: BaseRequest,
+        res: BaseResponse,
+        _: NormalisedURLPath,
+        __: HTTPMethod
+    ): Promise<boolean> => {
+        const options = {
+            config: this.config,
+            recipeId: this.getRecipeId(),
+            isInServerlessEnv: this.isInServerlessEnv,
+            recipeImplementation: this.recipeInterfaceImpl,
+            req,
+            res,
+            appInfo: this.getAppInfo(),
+        };
+        if (id === CREATE_DEVICE_API) {
+            return await createDeviceAPI(this.apiImpl, options);
+        } else if (id == VERIFY_DEVICE_API) {
+            return await verifyDeviceAPI(this.apiImpl, options);
+        } else if (id == VERIFY_CODE_API) {
+            return await verifyCodeAPI(this.apiImpl, options);
+        } else if (id == REMOVE_DEVICE_API) {
+            return await removeDeviceAPI(this.apiImpl, options);
+        } else if (id == LIST_DEVICE_API) {
+            return await listDevicesAPI(this.apiImpl, options);
+        }
+
+        return false;
+    };
+
+    handleError = async (err: STError, _: BaseRequest, res: BaseResponse): Promise<void> => {
+        if (err.fromRecipe === TotpRecipe.RECIPE_ID) {
+            if (err.type === TotpError.TOTP_NOT_ENABLED_ERROR) {
+                sendNon200ResponseWithMessage(res, "TOTP is not enabled for the user", 400); // bad req
+            }
+        }
+        throw err;
+    };
+
+    getAllCORSHeaders = (): string[] => {
+        return [];
+    };
+
+    isErrorFromThisRecipe = (err: any): err is STError => {
+        return STError.isErrorFromSuperTokens(err) && err.fromRecipe === TotpRecipe.RECIPE_ID;
+    };
+
+    getEmailOrPhoneForRecipeUserId: GetEmailOrPhoneForRecipeUserIdFunc = async (recipeUserId, userContext) => {
+        if (this.config.getEmailOrPhoneForRecipeUserId !== undefined) {
+            const userRes = await this.config.getEmailOrPhoneForRecipeUserId(recipeUserId, userContext);
+            if (userRes.status !== "UNKNOWN_USER_ID_ERROR") {
+                return userRes;
+            }
+        }
+
+        let user = await getUser(recipeUserId.getAsString(), userContext);
+
+        if (user === undefined) {
+            return {
+                status: "UNKNOWN_USER_ID_ERROR",
+            };
+        }
+
+        const primaryLoginMethod = user.loginMethods.find((method) => method.recipeUserId.getAsString() === user!.id);
+
+        if (primaryLoginMethod === undefined) {
+            return {
+                status: "UNKNOWN_USER_ID_ERROR",
+            };
+        }
```

Comment:
This is not true. There can be other login methods even if the primary user's login method is removed. If this is undefined, then you can " // fallback on trying the first email or phone number" method